processor/src/main/java/org/mapstruct/ap/internal/model/source/MappingOptions.java

```diff
@@ -54,13 +58,171 @@ public MappingOptions(Map<String, List<Mapping>> mappings, IterableMapping itera
+            }
+        }
+        return found;
+    }
+
+    /**
+     *
+     * @return all dependencies to other properties the contained mappings are dependent on
+     */
+    public List<String> collectNestedDependsOn() {
+
+        List<String> nestedDependsOn = new ArrayList<String>();
+        for ( List<Mapping> mappingList : mappings.values() ) {
+            for ( Mapping mapping : mappingList ) {
+                nestedDependsOn.addAll( mapping.getDependsOn() );
+            }
+        }
+        return nestedDependsOn;
+    }
+
+
+    /**
+     * Splits the MappingOptions into possibly more MappingOptions based on each source method parameter type.
+     *
+     * Note: this method is used for forging nested update methods. For that purpose, the same method with all
+     *       joined mappings should be generated. See also: NestedTargetPropertiesTest#shouldMapNestedComposedTarget
+     *
+     * @return the split mapping options.
+     *
+     */
+    public Map<Parameter, MappingOptions> splitOnSourceParameterType() {
+
+        Set<Parameter> parameters = new HashSet<Parameter>();
+        Map<Type, List<Mapping>> mappingsKeyedByParameterType = new HashMap<Type, List<Mapping>>();
+        for ( List<Mapping> mappingList : mappings.values() ) {
+            for ( Mapping mapping : mappingList ) {
+                if ( mapping.getSourceReference() != null && mapping.getSourceReference().isValid() ) {
+                    Parameter parameter = mapping.getSourceReference().getParameter();
+                    parameters.add( parameter );
+                    if ( !mappingsKeyedByParameterType.containsKey( parameter.getType() ) ) {
+                        mappingsKeyedByParameterType.put( parameter.getType(), new ArrayList<Mapping>() );
+                    }
+                    mappingsKeyedByParameterType.get( parameter.getType() ).add( mapping );
+                }
+            }
+        }
+
+        Map<Parameter, MappingOptions> result = new HashMap<Parameter, MappingOptions>();
+        for ( Parameter parameter : parameters ) {
+            List<Mapping> mappingsPerParameter = mappingsKeyedByParameterType.get( parameter.getType() );
+            result.put( parameter, MappingOptions.forMappingsOnly( sortByTargetName( mappingsPerParameter ) ) );
+        }
+        return result;
+    }
+
+    private Map<String, List<Mapping>> sortByTargetName( List<Mapping> mappingList ) {
+        Map<String, List<Mapping>> result = new HashMap<String, List<Mapping>>();
+        for ( Mapping mapping : mappingList ) {
+            if ( !result.containsKey( mapping.getTargetName() ) ) {
+                result.put( mapping.getTargetName(), new ArrayList<Mapping>() );
+            }
+            result.get( mapping.getTargetName() ).add( mapping );
+        }
+        return result;
+    }
+
+    public void reInit( Parameter sourceParameter ) {
```

Comment:
we place the `Mapping`s on top of a new method. So we need to associate those with the new source parameter in this method. There can be only one source parameter in generated forged methods... Something along those lines?